src/test/AludelV3.t.sol

```diff
@@ -0,0 +1,304 @@
+
+        data = aludel.getAludelData();
+        assertEq(data.rewardSharesOutstanding, REWARD_AMOUNT * BASE_SHARES_PER_WEI * 5);
+        // previous four schedule amounts and half of the fifth.
+        assertEq(
+            aludel.calculateUnlockedRewards(block.timestamp),
+            REWARD_AMOUNT * 4 + REWARD_AMOUNT / 2
+        );
+        // fifth schedule shares are half-locked
+        assertEq(
+            aludel.calculateSharesLocked(data.rewardSchedules, block.timestamp),
+            REWARD_AMOUNT * BASE_SHARES_PER_WEI / 2
+        );
+    }
+
+    function test_stakes_no_amount_staked() public {
+        Crucible crucible = Utils.createCrucible(user, crucibleFactory);
+
+        vm.warp(block.timestamp + 15);
+        bytes memory lockSig = Utils.getLockPermission(
+            user, crucible, address(aludel), stakingToken, 0
+        );
+        vm.expectRevert(AludelV3.NoAmountStaked.selector);
+        aludel.stake(address(crucible), 0, lockSig);    
+    }
+
+    function test_stakes_invalid_vault() public {
+        
+        // Create a crucible instance from a different factory
+        Crucible crucibleTemplate = new Crucible();
+        CrucibleFactory crucibleFactory = new CrucibleFactory(address(crucibleTemplate));
+        Crucible crucible = Utils.createCrucible(user, crucibleFactory);
+
+        Utils.fundMockToken(address(crucible), stakingToken, STAKE_AMOUNT);
+
+        vm.warp(block.timestamp + 15);
+        bytes memory lockSig = Utils.getLockPermission(
+            user, crucible, address(aludel), stakingToken, 1
+        );
+        vm.expectRevert(AludelV3.InvalidVault.selector);
+        aludel.stake(address(crucible), 1, lockSig); 
+    }
+
+    function test_stakes_max_stakes_reached() public {
+
+        Crucible crucible = Utils.createCrucible(user, crucibleFactory);
+        Utils.fundMockToken(address(crucible), stakingToken, STAKE_AMOUNT);
+        
+        bytes memory lockSig;
+
+        // stake 30 times 1 wei
+        for (uint i = 0; i < 30; i++) {
+            Utils.stake(user, crucible, aludel, stakingToken, 1);
+        }
+
+        // 31th stake should revert
+        lockSig = Utils.getLockPermission(user, crucible, address(aludel), stakingToken, 1);
+        vm.expectRevert(AludelV3.MaxStakesReached.selector);
+        aludel.stake(address(crucible), 1, lockSig);
+    }
+
+    function test_aludel_stake_not_enough_balance() public {
+
+        Crucible crucible = Utils.createCrucible(user, crucibleFactory);
+        Utils.fundMockToken(address(crucible), stakingToken, STAKE_AMOUNT);
+        vm.warp(block.timestamp + 15);
+        bytes memory lockSig = Utils.getLockPermission(
+            user, crucible, address(aludel), stakingToken, STAKE_AMOUNT + 1
+        );
+        vm.expectRevert(bytes("UniversalVault: insufficient balance"));
+        aludel.stake(address(crucible), STAKE_AMOUNT + 1, lockSig);
+    }
+
+    function test_aludel_stake_invalid_permission() public {
+
```

Comment:
I wasn't gona comment on this, but since there's a duplicated line that should be removed below, I'd say you also remove whitespace right before the body of a function, since it doesn't serve any purpose IMO